Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean

```diff
@@ -0,0 +1,995 @@
+  simp_rw [Algebra.norm_eq_matrix_det <| CoordinateRing.basis W, Matrix.det_fin_two,
+    Algebra.leftMulMatrix_eq_repr_mul, basis_zero, mul_one, basis_one, smul_basis_mul_Y, map_add,
+    Finsupp.add_apply, map_smul, Finsupp.smul_apply, ← basis_zero, ← basis_one,
+    Basis.repr_self_apply, if_pos, if_neg, smul_eq_mul]
+  ring1
+#align weierstrass_curve.coordinate_ring.norm_smul_basis WeierstrassCurve.CoordinateRing.norm_smul_basis
+
+lemma coe_norm_smul_basis (p q : R[X]) :
+    ↑(Algebra.norm R[X] <| p • (1 : W.CoordinateRing) + q • mk W Y) =
+      mk W ((C p + C q * X) * (C p + C q * (-Y - C (C W.a₁ * X + C W.a₃)))) :=
+  AdjoinRoot.mk_eq_mk.mpr
+    ⟨C q ^ 2, by simp only [norm_smul_basis, WeierstrassCurve.polynomial]; C_simp; ring1⟩
+#align weierstrass_curve.coordinate_ring.coe_norm_smul_basis WeierstrassCurve.CoordinateRing.coe_norm_smul_basis
+
+-- porting note: BUG `cases` tactic does not modify assumptions
+lemma degree_norm_smul_basis [IsDomain R] (p q : R[X]) :
+    (Algebra.norm R[X] <| p • (1 : W.CoordinateRing) + q • mk W Y).degree =
+      max (2 • p.degree) (2 • q.degree + 3) := by
+  have hdp : (p ^ 2).degree = 2 • p.degree := degree_pow p 2
+  have hdpq : (p * q * (C W.a₁ * X + C W.a₃)).degree ≤ p.degree + q.degree + 1 := by
+    simpa only [degree_mul] using add_le_add_left degree_linear_le (p.degree + q.degree)
+  have hdq :
+      (q ^ 2 * (X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆)).degree = 2 • q.degree + 3 := by
+    rw [degree_mul, degree_pow, ← one_mul <| X ^ 3, ← C_1, degree_cubic <| one_ne_zero' R]
+  rw [norm_smul_basis]
+  by_cases hp : p = 0
+  · simpa only [hp, hdq, neg_zero, zero_sub, zero_mul, zero_pow zero_lt_two, degree_neg] using
+      (max_bot_left _).symm
+  · by_cases hq : q = 0
+    · simpa only [hq, hdp, sub_zero, zero_mul, mul_zero, zero_pow zero_lt_two] using
+        (max_bot_right _).symm
+    · rw [← not_congr degree_eq_bot] at hp hq
+      rcases hp' : p.degree with _ | dp
+      · exact (hp hp').elim
+      · rw [hp'] at hdp hdpq
+        rcases hq' : q.degree with _ | dq
+        · exact (hq hq').elim
+        · rw [hq'] at hdpq hdq
+          rcases le_or_lt dp (dq + 1) with hpq | hpq
+          · convert (degree_sub_eq_right_of_degree_lt <| (degree_sub_le _ _).trans_lt <|
+                      max_lt_iff.mpr ⟨hdp.trans_lt _, hdpq.trans_lt _⟩).trans
+              (max_eq_right_of_lt _).symm <;> rw [hdq] <;>
+              exact WithBot.coe_lt_coe.mpr <| by linarith only [hpq]
+          · rw [sub_sub]
+            convert (degree_sub_eq_left_of_degree_lt <| (degree_add_le _ _).trans_lt <|
+                      max_lt_iff.mpr ⟨hdpq.trans_lt _, hdq.trans_lt _⟩).trans
+              (max_eq_left_of_lt _).symm <;> rw [hdp] <;>
+              exact WithBot.coe_lt_coe.mpr <| by linarith only [hpq]
+#align weierstrass_curve.coordinate_ring.degree_norm_smul_basis WeierstrassCurve.CoordinateRing.degree_norm_smul_basis
+
+variable {W}
+
+-- porting note: replaced `dec_trivial` with explicit lemma
```

Comment:
Again, please move to the exact location of the regression, so it's obvious without opening mathlib3 where one should hope to write `dec_trivial`.